pkg/virt-config/configuration_test.go

```diff
@@ -0,0 +1,630 @@
+	},
+		Entry("is positive, GetMemBalloonStatsPeriod should return period", &validMemBalloonStatsPeriod, uint32(3)),
+		Entry("when unset, GetMemBalloonStatsPeriod should return 10", nil, uint32(10)),
+	)
+
+	DescribeTable(" when useEmulation", func(value bool, result bool) {
+		clusterConfig, _, _ := testutils.NewFakeClusterConfigUsingKVConfig(&v1.KubeVirtConfiguration{
+			DeveloperConfiguration: &v1.DeveloperConfiguration{
+				UseEmulation: value,
+			},
+		})
+		Expect(clusterConfig.AllowEmulation()).To(Equal(result))
+	},
+		Entry("is true, AllowEmulation should return true", true, true),
+		Entry("is false, AllowEmulation should return false", false, false),
+	)
+
+	trueValue := true
+	falseValue := false
+	DescribeTable(" when permitSlirpInterface", func(value *bool, result bool) {
+		clusterConfig, _, _ := testutils.NewFakeClusterConfigUsingKVConfig(&v1.KubeVirtConfiguration{
+			NetworkConfiguration: &v1.NetworkConfiguration{
+				PermitSlirpInterface: value,
+			},
+		})
+
+		Expect(clusterConfig.IsSlirpInterfaceEnabled()).To(Equal(result))
+	},
+		Entry("is true, IsSlirpInterfaceEnabled should return true", &trueValue, true),
+		Entry("is false, IsSlirpInterfaceEnabled should return false", &falseValue, false),
+		Entry("when unset, IsSlirpInterfaceEnabled should return false", nil, false),
+	)
+
+	DescribeTable(" when permitBridgeInterfaceOnPodNetwork", func(value *bool, result bool) {
+		clusterConfig, _, _ := testutils.NewFakeClusterConfigUsingKVConfig(&v1.KubeVirtConfiguration{
+			NetworkConfiguration: &v1.NetworkConfiguration{
+				PermitBridgeInterfaceOnPodNetwork: value,
+			},
+		})
+
+		Expect(clusterConfig.IsBridgeInterfaceOnPodNetworkEnabled()).To(Equal(result))
+	},
+		Entry("is true, IsBridgeInterfaceOnPodNetworkEnabled should return true", &trueValue, true),
+		Entry("is false, IsBridgeInterfaceOnPodNetworkEnabled should return false", &falseValue, false),
+		Entry("when unset, IsBridgeInterfaceOnPodNetworkEnabled should return true", nil, true),
+	)
+
+	DescribeTable(" when defaultNetworkInterface", func(value string, result string) {
+		clusterConfig, _, _ := testutils.NewFakeClusterConfigUsingKVConfig(&v1.KubeVirtConfiguration{
+			NetworkConfiguration: &v1.NetworkConfiguration{
+				NetworkInterface: value,
+			},
+		})
+		Expect(clusterConfig.GetDefaultNetworkInterface()).To(Equal(result))
+	},
+		Entry("is bridge, GetDefaultNetworkInterface should return bridge", "bridge", "bridge"),
+		Entry("is slirp, GetDefaultNetworkInterface should return slirp", "slirp", "slirp"),
+		Entry("is masquerade, GetDefaultNetworkInterface should return masquerade", "masquerade", "masquerade"),
+		Entry("when unset, GetDefaultNetworkInterface should return the default", "", "bridge"),
+		// todo: is it a bug? this use case returns "invalid" instead of the default "bridge"
+		//Entry("when invalid, GetDefaultNetworkInterface should return the default", "invalid", "bridge"),
```

Comment:
For the case that someone manages to inject bad values into the config, the idea was to stick with defaults (or the last valid configuration). You probably removed some of the value validation code together with the config map removal.

Reply:
@rmohr - the removed code was part of the CM processing. Most of it took string values, validated them and converted them into the right final value in KV config. There was no such code for configuration from the KubeVirt CR, and it can't be here anyway - it's too late, because if we're here, then the KubeVirt CR is already updated in etcd.

This should be solved in the CRD or a validation webhookl in virt-api, and I think it's out of scope of this PR.

Reply:
Still, as we discussed on slack, it should have been there, like for the configmap. You are right that at this stage validation can't prevent the data from being persisted, but our components should refuse to pick up configs which they can't understand and cpnti with the last config they could understand.

Reply:
Understood. However, what if one of the pods will restarted? If I understand correctly, it will use the default configuation, while the other pods (not even the same KV application) will use an old configuration, and then we have a problematic situation when some pods use default and some use some old configuration. 

So I agree that we need to add validation in virt-config, but I would also prevent this situation by reject wrong values on the API level - in the CRD if possible.

Reply:
Will something like this work? https://github.com/kubevirt/api/pull/1

Reply:
I think that there are always reasons why one can end up with a config which is not correct from the components view. Therefore we need to keep the last working one.

You are right that the case where we get an invalid config after a restart and we fall back to the default config is a little bit more debate-able and I think you are right that it deserves extra handling. One option would be really during the start-up, if we we can't get an initial valid config, we simply refuse to start. This way we keep the following properties:

 * New config fields are not an issue, since old components can't even see them at this stage (the desired behaviour during upgrades)
  * If one attempts a downgrade without "downgrading" the config first, old components refuse to start
  * If an invalid config gets injected by accident, we don't immediately bring down kubevirt and there is a short window which allows admins to restore the installation without causing a huge outage.

What do you think? Other ideas how to handle this special case? Maybe a read-only mode in that special case of startup without a valid config?